src/Compilers/Core/SharedCollections/ArrayBuilder.cs

```diff
@@ -416,5 +417,41 @@ public void AddMany(T item, int count)
                 Add(item);
             }
         }
+
+        public void RemoveDuplicates()
+        {
+            var set = PooledHashSet<T>.GetInstance();
+
+            int j = 0;
+            for (int i = 0; i < Count; i++)
+            {
+                if (set.Add(this[i]))
+                {
+                    j++;
+                }
+
+                this[j] = this[i];
+            }
+
+            set.Free();
+        }
+
+        public ImmutableArray<S> SelectDistinct<S>(Func<T, S> selector)
+        {
+            var result = ArrayBuilder<S>.GetInstance();
```

Comment:
We could, but in general we don't know how many distinct items are there gonna be.